File: foundational_design_patterns/7_multi_agent_collaboration/src/orchestrator_worker.py
"""
Multi-Agent Collaboration: Orchestrator-Worker topology

Anchor scenario: deep research triage for support operations.

Preferred runtime in 2026:
- langgraph-supervisor (if installed)
Fallback:
- lightweight LangGraph orchestrator with explicit worker roles
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass, field

from pathlib import Path

ROOT_DIR = next(parent for parent in Path(__file__).resolve().parents if (parent / "ssl_fix.py").exists())
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from repo_support import configure_example, get_default_model

configure_example(__file__)

from langchain_openai import ChatOpenAI
from langgraph.graph import END, StateGraph


@dataclass
class OrchestratorState:
    goal: str
    work_items: list[str] = field(default_factory=list)
    results: list[str] = field(default_factory=list)
    final_report: str = ""


llm = ChatOpenAI(model=get_default_model(), temperature=0)


def supervisor_plan(state: OrchestratorState) -> OrchestratorState:
    prompt = f"""
Break this support-ops research goal into exactly 3 worker tasks:
{state.goal}

Return one task per line.
"""
    tasks = [line.strip("-• ").strip() for line in llm.invoke(prompt).content.splitlines() if line.strip()]
    state.work_items = tasks[:3]
    return state


def worker_research(state: OrchestratorState) -> OrchestratorState:
    if not state.work_items:
        return state

    task = state.work_items.pop(0)
    result = llm.invoke(
        f"Research this subtask for support operations and return 2 concise bullets:\n{task}"
    ).content
    state.results.append(f"TASK: {task}\n{result}")
    return state


def synthesize(state: OrchestratorState) -> OrchestratorState:
    prompt = f"""
Synthesize a final report from these worker outputs.

Goal: {state.goal}
Worker outputs:
{state.results}

Return:
- executive summary
- immediate actions (3 bullets)
- escalation recommendation
"""
    state.final_report = llm.invoke(prompt).content
    return state


def route_worker(state: OrchestratorState) -> str:
    return "worker" if state.work_items else "synthesize"


def try_supervisor_package() -> bool:
    try:
        import langgraph_supervisor  # noqa: F401

        print("✅ langgraph-supervisor detected. (This demo uses lightweight fallback graph for portability.)")
        return True
    except Exception:
        print("ℹ️ langgraph-supervisor not installed; using built-in fallback graph.")
        return False


def build_graph():
    graph = StateGraph(OrchestratorState)
    graph.add_node("plan", supervisor_plan)
    graph.add_node("worker", worker_research)
    graph.add_node("synthesize", synthesize)

    graph.set_entry_point("plan")
    graph.add_conditional_edges("plan", route_worker, {"worker": "worker", "synthesize": "synthesize"})
    graph.add_conditional_edges("worker", route_worker, {"worker": "worker", "synthesize": "synthesize"})
    graph.add_edge("synthesize", END)
    return graph.compile()


def main() -> None:
    print("\n" + "=" * 80)
    print("MULTI-AGENT — ORCHESTRATOR/WORKER")
    print("=" * 80)

    if not os.getenv("OPENAI_API_KEY"):
        print("❌ OPENAI_API_KEY is required.")
        return

    try_supervisor_package()

    app = build_graph()
    goal = (
        "Investigate enterprise customer complaints about payment approval latency and produce "
        "a triage recommendation for support operations."
    )
    out = app.invoke(OrchestratorState(goal=goal))

    if isinstance(out, dict):
        final_report = out.get("final_report", "")
    else:
        final_report = getattr(out, "final_report", "")

    print("\n🧾 Final report:")
    print(final_report)


if __name__ == "__main__":
    main()

File: foundational_design_patterns/7_multi_agent_collaboration/typescript/src/orchestrator_worker.ts
/**
 * Multi-Agent Collaboration: Orchestrator-Worker topology
 * — TypeScript port of src/orchestrator_worker.py.
 */

import { ChatOpenAI } from "@langchain/openai";
import { Annotation, END, StateGraph } from "@langchain/langgraph";

const MODEL = process.env.OPENAI_MODEL ?? "gpt-4o-mini";

export const OrchestratorState = Annotation.Root({
  goal: Annotation<string>({
    reducer: (_left, right) => right,
    default: () => "",
  }),
  work_items: Annotation<string[]>({
    reducer: (_left, right) => right,
    default: () => [],
  }),
  results: Annotation<string[]>({
    reducer: (left, right) => left.concat(right),
    default: () => [],
  }),
  final_report: Annotation<string>({
    reducer: (_left, right) => right,
    default: () => "",
  }),
});

type OrchestratorStateType = typeof OrchestratorState.State;

const llm = new ChatOpenAI({
  model: MODEL,
  temperature: 0,
});

export async function supervisorPlan(
  state: OrchestratorStateType,
): Promise<Partial<OrchestratorStateType>> {
  const prompt = `
Break this support-ops research goal into exactly 3 worker tasks:
${state.goal}

Return one task per line.
`;

  const response = await llm.invoke(prompt);
  const tasks = String(response.content ?? "")
    .split("\n")
    .map((line) => line.trim().replace(/^[-•]\s*/, ""))
    .filter(Boolean)
    .slice(0, 3);

  return { work_items: tasks };
}

export async function workerResearch(
  state: OrchestratorStateType,
): Promise<Partial<OrchestratorStateType>> {
  if (!state.work_items.length) {
    return {};
  }

  const [task, ...remaining] = state.work_items;
  const response = await llm.invoke(
    `Research this subtask for support operations and return 2 concise bullets:\n${task}`,
  );

  return {
    work_items: remaining,
    results: [`TASK: ${task}\n${String(response.content ?? "")}`],
  };
}

export async function synthesize(
  state: OrchestratorStateType,
): Promise<Partial<OrchestratorStateType>> {
  const prompt = `
Synthesize a final report from these worker outputs.

Goal: ${state.goal}
Worker outputs:
${state.results.join("\n")}

Return:
- executive summary
- immediate actions (3 bullets)
- escalation recommendation
`;

  const response = await llm.invoke(prompt);
  return { final_report: String(response.content ?? "") };
}

export function routeWorker(state: OrchestratorStateType): string {
  return state.work_items.length ? "worker" : "synthesize";
}

export async function trySupervisorPackage(): Promise<boolean> {
  try {
    const importer = new Function(
      "specifier",
      "return import(specifier);",
    ) as (specifier: string) => Promise<unknown>;
    await importer("langgraph-supervisor");
    console.log(
      "✅ langgraph-supervisor detected. (This demo uses lightweight fallback graph for portability.)",
    );
    return true;
  } catch {
    console.log(
      "ℹ️ langgraph-supervisor not installed; using built-in fallback graph.",
    );
    return false;
  }
}

export function buildGraph() {
  return new StateGraph(OrchestratorState)
    .addNode("plan", supervisorPlan)
    .addNode("worker", workerResearch)
    .addNode("synthesize", synthesize)
    .addEdge("__start__", "plan")
    .addConditionalEdges("plan", routeWorker, {
      worker: "worker",
      synthesize: "synthesize",
    })
    .addConditionalEdges("worker", routeWorker, {
      worker: "worker",
      synthesize: "synthesize",
    })
    .addEdge("synthesize", END)
    .compile();
}

async function main(): Promise<void> {
  console.log("\n" + "=".repeat(80));
  console.log("MULTI-AGENT — ORCHESTRATOR/WORKER");
  console.log("=".repeat(80));

  if (!process.env.OPENAI_API_KEY) {
    console.log("❌ OPENAI_API_KEY is required.");
    return;
  }

  await trySupervisorPackage();

  const app = buildGraph();
  const goal =
    "Investigate enterprise customer complaints about payment approval latency and produce " +
    "a triage recommendation for support operations.";
  const output = await app.invoke({
    goal,
    work_items: [],
    results: [],
    final_report: "",
  });

  console.log("\n🧾 Final report:");
  console.log(output.final_report);
}

if (import.meta.main) {
  await main();
}
